### ISSUE TITLE
Camera.get_output_resolution_matrix() is surprisingly memory hungry

### ISSUE THREAD

**Issue by sbailey** [desihub/specsim#43]
Camera.get_output_resolution_matrix() is surprisingly memory hungry, temporarily requiring several GB of memory even though the output resolution matrix is a sparse matrix.  This is problematic for Travis tests which have a memory limit, and will be problematic for future NERSC machines that have less memory per core.  Please review if there is way to generate the resolution matrix while using less memory.  This is a blocking factor for desisim Travis tests that use specsim.

See desihub/desisim#145 for the background to this issue and a specific example.

**Comment by dkirkby** [comment]
The problematic `get_output_resolution_matrix()` function is not used by specsim and only provided for convenience, but both quickbrick and quickgen are calling it to fill the resolution matrix arrays in the output brick file. I think the solution here is to update `desisim` to replace this line from quickbrick (and a similar one from quickgen):
```
R = Resolution(camera.get_output_resolution_matrix())
```
with new code that can directly translate from the sparse representation used internally by specsim (scipy.sparse CSR format) to the sparse representation used in the brick file, without going through an intermediate dense matrix.  The solution could involve changes to desisim and/or specsim.

**Comment by sbailey** [comment]
Thanks for looking into this.  I think this could be accomplished via
* `diagdata = camera._resolution_matrix.todia().data`
* downsample `diagdata` taking care of alignment, edge effects, etc if `diagdata.shape` isn't evenly divisible by the downsampling size.  This is the tricky part.
* recast the downsampled `diagdata` into a `scipy.sparse.dia_matrix((diags, offsets), shape)`

Updating this from the desisim side would require access to the hidden variable `Camera._resolution_matrix` and potentially tricky downsampling logic given that that array isn't necessarily evenly divisible by the downsampling size, and the number of non-zero diagonal elements also aren't necessarily evenly divisible by the downsampling size.

I suggest that `get_output_resolution_matrix()` should be updated to downsample sparse -> sparse without going through a dense matrix, and return the downsampled sparse matrix instead of a dense matrix.  specsim could then get the desired resolution matrix output data via `camera.get_output_resolution_matrix().todia().data`, while still keeping the downsampling logic in a single testable place (`Camera.get_output_resolution_matrix()`).

**Comment by dkirkby** [comment]
I have re-implemented `get_output_resolution_matrix()` to perform the downsampling directly on the CSR sparse arrays used for the internal high-resolution matrix.  The results are not identical, presumably because of different round-off errors, but pass the `np.allclose()` test.  The memory requirements now scale with the number of non-zero elements in the downsampled matrix instead of the full size of the high-resolution matrix.  The timing has also improved:

| config | before | after |
|-------|--------|------|
| test    | 1.8s      | 0.3s  |
| desi   | 4.7s      | 1.0s   |

The factor of 3 between `test` and `desi` is because the test config only has one camera.  Is ~1.0s fast enough for the desisim travis tests?

**Comment by sbailey** [comment]
Thanks.  The Travis problem was the memory footprint killing off the test jobs, not the timing per-se.  This should enable the tests to fit within the 3 GB memory limits of Travis.  If not, I'll reopen the issue (it was auto-closed when I merged PR #50).

I had assumed that the slow timing of the quickgen/quickbrick tests was due to re-creating fresh copies of `specsim.simulator.Simulator('desi')` for every test to ensure leftover state from one test doesn't impact the next one.  But if `get_output_resolution_matrix()` was also taking 3-5 seconds per channel per test that would also significantly contribute to the slowness of the tests.  I (and/or @rstaten) can re-evaluate the speed of the desisim quickgen/quickbrick tests and open a new issue if there are other requests.
